// @angular/core/schematics/migrations/testbed-teardown/util.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/testbed-teardown/util", ["require", "exports", "typescript", "@angular/core/schematics/utils/typescript/imports", "@angular/core/schematics/utils/typescript/symbol"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.migrateTestModuleMetadataLiteral = exports.getInitTestEnvironmentLiteralReplacement = exports.findTestModuleMetadataNodes = exports.findInitTestEnvironmentCalls = void 0;
    const typescript_1 = __importDefault(require("typescript"));
    const imports_1 = require("@angular/core/schematics/utils/typescript/imports");
    const symbol_1 = require("@angular/core/schematics/utils/typescript/symbol");
    /** Finds the `initTestEnvironment` calls that need to be migrated. */
    function findInitTestEnvironmentCalls(typeChecker, allSourceFiles) {
        const callsToMigrate = new Set();
        let totalCalls = 0;
        allSourceFiles.forEach(sourceFile => {
            sourceFile.forEachChild(function walk(node) {
                if (typescript_1.default.isCallExpression(node) && typescript_1.default.isPropertyAccessExpression(node.expression) &&
                    typescript_1.default.isIdentifier(node.expression.name) &&
                    node.expression.name.text === 'initTestEnvironment' &&
                    isTestBedAccess(typeChecker, node.expression)) {
                    totalCalls++;
                    if (shouldMigrateInitTestEnvironment(node)) {
                        callsToMigrate.add(node);
                    }
                }
                node.forEachChild(walk);
            });
        });
        return {
            // Sort the nodes so that they will be migrated in reverse source order (nodes at the end of
            // the file are migrated first). This avoids issues where a migrated node will offset the
            // bounds of all nodes that come after it. Note that the nodes here are from all of the
            // passed in source files, but that doesn't matter since the later nodes will still appear
            // after the earlier ones.
            callsToMigrate: sortInReverseSourceOrder(Array.from(callsToMigrate)),
            totalCalls
        };
    }
    exports.findInitTestEnvironmentCalls = findInitTestEnvironmentCalls;
    /** Finds the `configureTestingModule` and `withModule` calls that need to be migrated. */
    function findTestModuleMetadataNodes(typeChecker, sourceFile) {
        const testModuleMetadataLiterals = new Set();
        const withModuleImport = (0, imports_1.getImportSpecifier)(sourceFile, '@angular/core/testing', 'withModule');
        sourceFile.forEachChild(function walk(node) {
            if (typescript_1.default.isCallExpression(node)) {
                const isConfigureTestingModuleCall = typescript_1.default.isPropertyAccessExpression(node.expression) &&
                    typescript_1.default.isIdentifier(node.expression.name) &&
                    node.expression.name.text === 'configureTestingModule' &&
                    isTestBedAccess(typeChecker, node.expression) && shouldMigrateModuleConfigCall(node);
                const isWithModuleCall = withModuleImport && typescript_1.default.isIdentifier(node.expression) &&
                    (0, symbol_1.isReferenceToImport)(typeChecker, node.expression, withModuleImport) &&
                    shouldMigrateModuleConfigCall(node);
                if (isConfigureTestingModuleCall || isWithModuleCall) {
                    testModuleMetadataLiterals.add(node.arguments[0]);
                }
            }
            node.forEachChild(walk);
        });
        // Sort the nodes so that they will be migrated in reverse source order (nodes at the end of
        // the file are migrated first). This avoids issues where a migrated node will offset the
        // bounds of all nodes that come after it.
        return sortInReverseSourceOrder(Array.from(testModuleMetadataLiterals));
    }
    exports.findTestModuleMetadataNodes = findTestModuleMetadataNodes;
    /**
     * Gets data that can be used to migrate a call to `TestBed.initTestEnvironment`.
     * The returned `span` is used to mark the text that should be replaced while the `text`
     * is the code that should be inserted instead.
     */
    function getInitTestEnvironmentLiteralReplacement(node, printer) {
        const literalProperties = [];
        const lastArg = node.arguments[node.arguments.length - 1];
        let span;
        let prefix;
        if (node.arguments.length > 2) {
            if (isFunction(lastArg)) {
                // If the last argument is a function, add the function as the `aotSummaries` property.
                literalProperties.push(typescript_1.default.createPropertyAssignment('aotSummaries', lastArg));
            }
            else if (typescript_1.default.isObjectLiteralExpression(lastArg)) {
                // If the property is an object literal, copy over all the properties.
                literalProperties.push(...lastArg.properties);
            }
            prefix = '';
            span = { start: lastArg.getStart(), end: lastArg.getEnd(), length: lastArg.getWidth() };
        }
        else {
            const start = lastArg.getEnd();
            prefix = ', ';
            span = { start, end: start, length: 0 };
        }
        // Finally push the teardown object so that it appears last.
        literalProperties.push(createTeardownAssignment());
        return {
            span,
            text: prefix +
                printer.printNode(typescript_1.default.EmitHint.Unspecified, typescript_1.default.createObjectLiteral(literalProperties, true), node.getSourceFile())
        };
    }
    exports.getInitTestEnvironmentLiteralReplacement = getInitTestEnvironmentLiteralReplacement;
    /** Migrates an object literal that is passed into `configureTestingModule` or `withModule`. */
    function migrateTestModuleMetadataLiteral(node) {
        return typescript_1.default.createObjectLiteral([...node.properties, createTeardownAssignment()], node.properties.length > 0);
    }
    exports.migrateTestModuleMetadataLiteral = migrateTestModuleMetadataLiteral;
    /** Returns whether a property access points to `TestBed`. */
    function isTestBedAccess(typeChecker, node) {
        var _a, _b;
        const symbolName = (_b = (_a = typeChecker.getTypeAtLocation(node.expression)) === null || _a === void 0 ? void 0 : _a.getSymbol()) === null || _b === void 0 ? void 0 : _b.getName();
        return symbolName === 'TestBed' || symbolName === 'TestBedStatic';
    }
    /** Whether a call to `initTestEnvironment` should be migrated. */
    function shouldMigrateInitTestEnvironment(node) {
        // If there is no third argument, we definitely have to migrate it.
        if (node.arguments.length === 2) {
            return true;
        }
        // This is technically a type error so we shouldn't mess with it.
        if (node.arguments.length < 2) {
            return false;
        }
        // Otherwise we need to figure out if the `teardown` flag is set on the last argument.
        const lastArg = node.arguments[2];
        // Note: the checks below will identify something like `initTestEnvironment(..., ..., {})`,
        // but they'll ignore a variable being passed in as the last argument like `const config = {};
        // initTestEnvironment(..., ..., config)`. While we can resolve the variable to its declaration
        // using `typeChecker.getTypeAtLocation(lastArg).getSymbol()?.valueDeclaration`, we deliberately
        // don't, because it introduces some complexity and we may end up breaking user code. E.g.
        // the `config` from the example above may be passed in to other functions or the `teardown`
        // flag could be added later on by a function call.
        // If the argument is an object literal and there are no
        // properties called `teardown`, we have to migrate it.
        if (isObjectLiteralWithoutTeardown(lastArg)) {
            return true;
        }
        // If the last argument is an `aotSummaries` function, we also have to migrate.
        if (isFunction(lastArg)) {
            return true;
        }
        // Otherwise don't migrate if we couldn't identify the last argument.
        return false;
    }
    /**
     * Whether a call to a module configuration function should be migrated. This covers
     * `TestBed.configureTestingModule` and `withModule` since they both accept `TestModuleMetadata`
     * as their first argument.
     */
    function shouldMigrateModuleConfigCall(node) {
        return node.arguments.length > 0 && isObjectLiteralWithoutTeardown(node.arguments[0]);
    }
    /** Returns whether a node is a function literal. */
    function isFunction(node) {
        return typescript_1.default.isArrowFunction(node) || typescript_1.default.isFunctionExpression(node) ||
            typescript_1.default.isFunctionDeclaration(node);
    }
    /** Checks whether a node is an object literal that doesn't contain a property called `teardown`. */
    function isObjectLiteralWithoutTeardown(node) {
        return typescript_1.default.isObjectLiteralExpression(node) && !node.properties.find(prop => {
            var _a;
            return ((_a = prop.name) === null || _a === void 0 ? void 0 : _a.getText()) === 'teardown';
        });
    }
    /** Creates a teardown configuration property assignment. */
    function createTeardownAssignment() {
        // `teardown: {destroyAfterEach: false}`
        return typescript_1.default.createPropertyAssignment('teardown', typescript_1.default.createObjectLiteral([typescript_1.default.createPropertyAssignment('destroyAfterEach', typescript_1.default.createFalse())]));
    }
    /** Sorts an array of AST nodes in reverse source order. */
    function sortInReverseSourceOrder(nodes) {
        return nodes.sort((a, b) => b.getEnd() - a.getEnd());
    }
});
