// @angular/core/schematics/utils/ng_component_template.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/utils/ng_component_template", ["require", "exports", "path", "typescript", "@angular/core/schematics/utils/line_mappings", "@angular/core/schematics/utils/ng_decorators", "@angular/core/schematics/utils/typescript/functions", "@angular/core/schematics/utils/typescript/property_name"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.NgComponentTemplateVisitor = void 0;
    const path_1 = require("path");
    const typescript_1 = __importDefault(require("typescript"));
    const line_mappings_1 = require("@angular/core/schematics/utils/line_mappings");
    const ng_decorators_1 = require("@angular/core/schematics/utils/ng_decorators");
    const functions_1 = require("@angular/core/schematics/utils/typescript/functions");
    const property_name_1 = require("@angular/core/schematics/utils/typescript/property_name");
    /**
     * Visitor that can be used to determine Angular templates referenced within given
     * TypeScript source files (inline templates or external referenced templates)
     */
    class NgComponentTemplateVisitor {
        constructor(typeChecker, _basePath, _tree) {
            this.typeChecker = typeChecker;
            this._basePath = _basePath;
            this._tree = _tree;
            this.resolvedTemplates = [];
        }
        visitNode(node) {
            if (node.kind === typescript_1.default.SyntaxKind.ClassDeclaration) {
                this.visitClassDeclaration(node);
            }
            typescript_1.default.forEachChild(node, n => this.visitNode(n));
        }
        visitClassDeclaration(node) {
            if (!node.decorators || !node.decorators.length) {
                return;
            }
            const ngDecorators = (0, ng_decorators_1.getAngularDecorators)(this.typeChecker, node.decorators);
            const componentDecorator = ngDecorators.find(dec => dec.name === 'Component');
            // In case no "@Component" decorator could be found on the current class, skip.
            if (!componentDecorator) {
                return;
            }
            const decoratorCall = componentDecorator.node.expression;
            // In case the component decorator call is not valid, skip this class declaration.
            if (decoratorCall.arguments.length !== 1) {
                return;
            }
            const componentMetadata = (0, functions_1.unwrapExpression)(decoratorCall.arguments[0]);
            // Ensure that the component metadata is an object literal expression.
            if (!typescript_1.default.isObjectLiteralExpression(componentMetadata)) {
                return;
            }
            const sourceFile = node.getSourceFile();
            const sourceFileName = sourceFile.fileName;
            // Walk through all component metadata properties and determine the referenced
            // HTML templates (either external or inline)
            componentMetadata.properties.forEach(property => {
                if (!typescript_1.default.isPropertyAssignment(property)) {
                    return;
                }
                const propertyName = (0, property_name_1.getPropertyNameText)(property.name);
                // In case there is an inline template specified, ensure that the value is statically
                // analyzable by checking if the initializer is a string literal-like node.
                if (propertyName === 'template' && typescript_1.default.isStringLiteralLike(property.initializer)) {
                    // Need to add an offset of one to the start because the template quotes are
                    // not part of the template content.
                    // The `getText()` method gives us the original raw text.
                    // We could have used the `text` property, but if the template is defined as a backtick
                    // string then the `text` property contains a "cooked" version of the string. Such cooked
                    // strings will have converted CRLF characters to only LF. This messes up string
                    // replacements in template migrations.
                    // The raw text returned by `getText()` includes the enclosing quotes so we change the
                    // `content` and `start` values accordingly.
                    const content = property.initializer.getText().slice(1, -1);
                    const start = property.initializer.getStart() + 1;
                    this.resolvedTemplates.push({
                        filePath: sourceFileName,
                        container: node,
                        content,
                        inline: true,
                        start: start,
                        getCharacterAndLineOfPosition: pos => typescript_1.default.getLineAndCharacterOfPosition(sourceFile, pos + start)
                    });
                }
                if (propertyName === 'templateUrl' && typescript_1.default.isStringLiteralLike(property.initializer)) {
                    const templateDiskPath = (0, path_1.resolve)((0, path_1.dirname)(sourceFileName), property.initializer.text);
                    // TODO(devversion): Remove this when the TypeScript compiler host is fully virtual
                    // relying on the devkit virtual tree and not dealing with disk paths. This is blocked on
                    // providing common utilities for schematics/migrations, given this is done in the
                    // Angular CDK already:
                    // https://github.com/angular/components/blob/3704400ee67e0190c9783e16367587489c803ebc/src/cdk/schematics/update-tool/utils/virtual-host.ts.
                    const templateDevkitPath = (0, path_1.relative)(this._basePath, templateDiskPath);
                    // In case the template does not exist in the file system, skip this
                    // external template.
                    if (!this._tree.exists(templateDevkitPath)) {
                        return;
                    }
                    const fileContent = this._tree.read(templateDevkitPath).toString();
                    const lineStartsMap = (0, line_mappings_1.computeLineStartsMap)(fileContent);
                    this.resolvedTemplates.push({
                        filePath: templateDiskPath,
                        container: node,
                        content: fileContent,
                        inline: false,
                        start: 0,
                        getCharacterAndLineOfPosition: pos => (0, line_mappings_1.getLineAndCharacterFromPosition)(lineStartsMap, pos),
                    });
                }
            });
        }
    }
    exports.NgComponentTemplateVisitor = NgComponentTemplateVisitor;
});
