// @angular/core/schematics/migrations/typed-forms/util.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/typed-forms/util", ["require", "exports", "typescript", "@angular/core/schematics/utils/typescript/imports", "@angular/core/schematics/utils/typescript/symbol"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.findFormBuilderCalls = exports.findControlClassUsages = exports.getAnyImport = exports.getFormBuilderImport = exports.getControlClassImports = exports.anySymbolName = exports.builderMethodNames = exports.controlClassNames = void 0;
    const typescript_1 = __importDefault(require("typescript"));
    const imports_1 = require("@angular/core/schematics/utils/typescript/imports");
    const symbol_1 = require("@angular/core/schematics/utils/typescript/symbol");
    exports.controlClassNames = ['AbstractControl', 'FormArray', 'FormControl', 'FormGroup'];
    exports.builderMethodNames = ['control', 'group', 'array'];
    exports.anySymbolName = 'AnyForUntypedForms';
    function getControlClassImports(sourceFile) {
        return exports.controlClassNames.map(cclass => (0, imports_1.getImportSpecifier)(sourceFile, '@angular/forms', cclass))
            .filter(v => v != null);
    }
    exports.getControlClassImports = getControlClassImports;
    function getFormBuilderImport(sourceFile) {
        return (0, imports_1.getImportSpecifier)(sourceFile, '@angular/forms', 'FormBuilder');
    }
    exports.getFormBuilderImport = getFormBuilderImport;
    function getAnyImport(sourceFile) {
        return (0, imports_1.getImportSpecifier)(sourceFile, '@angular/forms', exports.anySymbolName);
    }
    exports.getAnyImport = getAnyImport;
    function findControlClassUsages(sourceFile, typeChecker, importSpecifier) {
        var _a;
        if (importSpecifier === null)
            return [];
        let generic = `<${exports.anySymbolName}>`;
        if (importSpecifier.name.getText() === 'FormArray' ||
            ((_a = importSpecifier.propertyName) === null || _a === void 0 ? void 0 : _a.getText()) === 'FormArray') {
            generic = `<${exports.anySymbolName}[]>`;
        }
        const usages = [];
        const visitNode = (node) => {
            // Look for a `new` expression with no type arguments which references an import we care about:
            // `new FormControl()`
            if (typescript_1.default.isNewExpression(node) && !node.typeArguments &&
                (0, symbol_1.isReferenceToImport)(typeChecker, node.expression, importSpecifier)) {
                usages.push({ node: node.expression, generic });
            }
            typescript_1.default.forEachChild(node, visitNode);
        };
        typescript_1.default.forEachChild(sourceFile, visitNode);
        return usages;
    }
    exports.findControlClassUsages = findControlClassUsages;
    function findFormBuilderCalls(sourceFile, typeChecker, importSpecifier) {
        if (!importSpecifier)
            return [];
        const usages = new Array();
        typescript_1.default.forEachChild(sourceFile, function visitNode(node) {
            // Look for calls that look like `foo.<method to migrate>`.
            if (typescript_1.default.isCallExpression(node) && !node.typeArguments &&
                typescript_1.default.isPropertyAccessExpression(node.expression) && typescript_1.default.isIdentifier(node.expression.name) &&
                exports.builderMethodNames.includes(node.expression.name.text)) {
                const generic = node.expression.name.text === 'array' ? `<${exports.anySymbolName}[]>` : `<${exports.anySymbolName}>`;
                // Check whether the type of the object on which the function is called refers to the
                // provided import.
                if ((0, symbol_1.isReferenceToImport)(typeChecker, node.expression.expression, importSpecifier)) {
                    usages.push({ node: node.expression, generic });
                }
            }
            typescript_1.default.forEachChild(node, visitNode);
        });
        return usages;
    }
    exports.findFormBuilderCalls = findFormBuilderCalls;
});
